Add authenticating with IdentityServer4 (#432)
SenseNet/sensenet.github.io: _posts/2020-05-05-authentication-with-identity-server.md

---

We have changed to authentication for sensenet 7 to IdentityServer4. 
Here are the things you need to know when you use it in an single page application.

---

## What is IdentityServer4?

IdentityServer4 is an OpenID Connect and OAuth 2.0 framework for ASP.NET Core 3.0. Come again? What does that even mean? What is OpenID Connect and OAuth 2.0? 

Quoting from [https://oauth.net/2/](https://oauth.net/2/)
> OAuth 2.0 is the industry-standard protocol for authorization. OAuth 2.0 focuses on client developer simplicity while providing specific authorization flows for web applications, desktop applications, mobile phones, and living room devices. This specification and its extensions are being developed within the [IETF OAuth Working Group](https://www.ietf.org/mailman/listinfo/oauth).

Quoting from [https://openid.net/connect/](https://openid.net/connect/)
> OpenID Connect 1.0 is a simple identity layer on top of the OAuth 2.0 protocol. It allows Clients to verify the identity of the End-User based on the authentication performed by an Authorization Server, as well as to obtain basic profile information about the End-User in an interoperable and REST-like manner.

> Ok, so IdentityServer4 is implementing the industry-standard protocol for authorization. That sounds awesome. We don't have to implement hacky/non-standard ways to implement authorization, IdentityServer4 handles that for us.

Reviewer: ```suggestion
Ok, so IdentityServer4 is implementing the industry-standard protocol for authorization. That sounds awesome. We don't have to implement hacky/non-standard ways to implement authorization, IdentityServer4 does most of the work for us.
```

> ## How can I authenticate to sensenet from your SPA then?

Reviewer: ```suggestion
## How can I authenticate to sensenet from a SPA then?
```

> The easiest way is to use our newly created React component library for handling authentication through a provider ([@sensenet/authentication-oidc-react](https://www.npmjs.com/package/@sensenet/authentication-oidc-react)). It uses [oidc-client](https://www.npmjs.com/package/oidc-client) under the hood. If you are using another framework or just vanilla js, your best bet is this lib.

Reviewer: ```suggestion
The easiest way is to use our newly created React component library for handling authentication through a provider ([@sensenet/authentication-oidc-react](https://www.npmjs.com/package/@sensenet/authentication-oidc-react)). It uses [oidc-client](https://www.npmjs.com/package/oidc-client) under the hood. If you are using another framework or just vanilla js, this lib is your best bet.
```

- Add @sensenet/authentication-oidc-react to your react app either with `yarn add @sensenet/authentication-oidc-react` or with `npm install @sensenet/authentication-oidc-react`

> - Add a configuration file like this 

Reviewer: ```suggestion
- Add a configuration file like this:
```

import { RepositoryContext } from '@sensenet/hooks-react'
import React, { PropsWithChildren } from 'react'
import ReactDOM from 'react-dom'
import { BrowserRouter, useHistory } from 'react-router-dom'

ReactDOM.render(
  <AppProviders>
    <App />
  </AppProviders>,
  document.getElementById('root'),
)

export function AppProviders({ children }: PropsWithChildren<{}>) {
  return (
    <BrowserRouter>
      <AuthProvider>
        <RepositoryProvider>{children}</RepositoryProvider>
      </AuthProvider>
    </BrowserRouter>
  )
}

export const AuthProvider = ({ children }: PropsWithChildren<{}>) => {
  const history = useHistory()

  return (
    <AuthenticationProvider configuration={configuration} history={history}>
      {children}
    </AuthenticationProvider>
  )
}

export const RepositoryProvider = ({ children }: PropsWithChildren<{}>) => {
  const { oidcUser } = useOidcAuthentication()

  if (!oidcUser) {
    return <LoginForm />
  }

  return (
    <RepositoryContext.Provider value={new Repository({ repositoryUrl, token: oidcUser.access_token })}>
      {children}
    </RepositoryContext.Provider>
  )
}
```

- That is it! 🎉 You can now authenticate with your user.

> ### What if I don't know the authority url in the configuration?

Reviewer: ```suggestion
### What if I don't know the authority url?
```
